src/Compilers/CSharp/Portable/Binder/SafeContext.cs

```diff
@@ -0,0 +1,132 @@
+    /// - A *referent* is the variable being referenced by a `ref`.
+    /// - Informally, a variable is *live* if it has storage allocated for it (either on heap or stack).
+    /// - In this design, all SafeContexts have a known relationship to all other SafeContexts.
+    /// </remarks>
+    internal readonly struct SafeContext
+    {
+        private const uint CallingMethodRaw = 0;
+        private const uint ReturnOnlyRaw = 1;
+        private const uint CurrentMethodRaw = 2;
+
+        /// <summary>
+        /// For the purpose of escape verification we operate with the depth of local scopes.
+        /// The depth is a uint, with smaller number representing shallower/wider scopes.
+        /// Since sibling scopes do not intersect and a value cannot escape from one to another without
+        /// escaping to a wider scope, we can use simple depth numbering without ambiguity.
+        /// </summary>
+        private readonly uint _value;
+        private SafeContext(uint value) => _value = value;
+
+        /// <summary>
+        /// The "calling method" scope that is outside of the containing method/lambda.
+        /// If something can escape to this scope, it can escape to any scope in a given method through a ref parameter or return.
+        /// </summary>
+        public static readonly SafeContext CallingMethod = new SafeContext(CallingMethodRaw);
+
+        /// <summary>
+        /// The "return-only" scope that is outside of the containing method/lambda.
+        /// If something can escape to this scope, it can escape to any scope in a given method or can be returned, but it can't escape through a ref parameter.
+        /// </summary>
+        public static readonly SafeContext ReturnOnly = new SafeContext(ReturnOnlyRaw);
+
+        /// <summary>
+        /// The "current method" scope that is just inside the containing method/lambda.
+        /// If something can escape to this scope, it can escape to any scope in a given method, but cannot be returned.
+        /// </summary>
+        public static readonly SafeContext CurrentMethod = new SafeContext(CurrentMethodRaw);
+
+        /// <summary>
+        /// Gets a SafeContext which is "empty". i.e. which refers to a variable whose storage is never allocated.
+        /// </summary>
+        public static readonly SafeContext Empty = new SafeContext(uint.MaxValue);
+
+        /// <summary>
+        /// Gets a SafeContext which is narrower than the given SafeContext.
+        /// Used to "enter" a nested local scope.
+        /// </summary>
+        public SafeContext Narrower()
+        {
+            Debug.Assert(_value >= ReturnOnlyRaw);
+            return new SafeContext(_value + 1);
+        }
+
+        /// <summary>
+        /// Gets a SafeContext which is wider than the given SafeContext.
+        /// Used to "exit" a nested local scope.
+        /// </summary>
+        public SafeContext Wider()
+        {
+            Debug.Assert(_value >= CurrentMethodRaw);
```

Comment:
Why `CurrentMethodRaw vs. ReturnOnly`? It should be valid to get a wider scope from a `ReturnOnly` context.

Reply:
The purpose of the method is to "exit" a local scope, alongside `Narrower()` to "enter" a local scope. It's not intended to, for example, transition between the well-known "returnable" scopes. (We do allow this method to return `ReturnOnly`, though, to allow for uniformity when exiting the "top-level" block of a method in the visitor.)